File: server/lib/routes/analytics.dart
import 'dart:convert';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import '../services/analytics_service.dart';

Filters _extractFilters(Map<String, String> params) {
  return Filters(
    startDate: params['startDate'],
    endDate: params['endDate'],
    account: params['account'],
    category: params['category'],
  );
}

Router analyticsRoutes() {
  final router = Router();

  router.get('/api/analytics/categories', (Request request) {
    final filters = _extractFilters(request.url.queryParameters);
    return Response.ok(
      jsonEncode(getCategoryBreakdown(filters)),
      headers: {'Content-Type': 'application/json'},
    );
  });

  router.get('/api/analytics/monthly', (Request request) {
    final filters = _extractFilters(request.url.queryParameters);
    return Response.ok(
      jsonEncode(getMonthlyBreakdown(filters)),
      headers: {'Content-Type': 'application/json'},
    );
  });

  router.get('/api/analytics/weekly', (Request request) {
    final filters = _extractFilters(request.url.queryParameters);
    return Response.ok(
      jsonEncode(getWeeklyBreakdown(filters)),
      headers: {'Content-Type': 'application/json'},
    );
  });

  return router;
}

File: server/src/routes/analytics.ts
import { Hono } from 'hono';
import {
  getCategoryBreakdown,
  getMonthlyBreakdown,
  getWeeklyBreakdown,
} from '../services/analyticsService.js';
import type { AnalyticsFilters } from '../types.js';

const app = new Hono();

app.get('/api/analytics/categories', (c) => {
  const filters: AnalyticsFilters = {
    startDate: c.req.query('startDate'),
    endDate: c.req.query('endDate'),
    account: c.req.query('account'),
    category: c.req.query('category'),
  };
  return c.json(getCategoryBreakdown(filters));
});

app.get('/api/analytics/monthly', (c) => {
  const filters: AnalyticsFilters = {
    startDate: c.req.query('startDate'),
    endDate: c.req.query('endDate'),
    account: c.req.query('account'),
    category: c.req.query('category'),
  };
  return c.json(getMonthlyBreakdown(filters));
});

app.get('/api/analytics/weekly', (c) => {
  const filters: AnalyticsFilters = {
    startDate: c.req.query('startDate'),
    endDate: c.req.query('endDate'),
    account: c.req.query('account'),
    category: c.req.query('category'),
  };
  return c.json(getWeeklyBreakdown(filters));
});

export default app;
